man/man3/v470_utils.3
\" -*- nroff -*-

\" v450_utils -- V450 library interface
\"
\" This program is free software; you can redistribute it and/or modify
\" it under the terms of the GNU General Public License as published by
\" the Free Software Foundation; either version 2 of the License, or (at
\" your option) any later version.
\"
\" This program is distributed in the hope that it will be useful, but
\" WITHOUT ANY WARRANTY; without even the implied warranty of
\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
\" General Public License for more details.
\"
\" You should have received a copy of the GNU General Public License
\" along with this program. If not, see <http://www.gnu.org/licenses/>.
\"
\" Author: Dean W. Anneser
\" Company: RTLinux Solutions LLC for Highland Technology, Inc.
\" Date: Mon Aug 09 09:10:09 2021

.TH v470_utils 3 "user level utility library for VME V470 card"

.SH NAME
.nf
\f4v470_utils\f1 -- collection of routines to interface with the VME V470 16 channel D/A card.
.fi

.SH SYNOPSIS
.SS INCLUDE FILES
\f4#include \f2"v470_utils_proto.h"\f1
.br
\f4#include \f2"v470_regs_sdef.h"\f1
.br

.SS READ/WRITE REGISTERS
\f4unsigned short v470_read_vximfr(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_read_vxitype(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_read_serial(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_read_romid(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_read_romrev(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_read_cflags(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_read_rflags(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_read_relays(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_write_relays(V470_REGS *\f2pb\f4, unsigned short \f2relays\f4)\f1
.br
\f4unsigned short v470_read_uled(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_write_uled(V470_REGS *\f2pb\f4, unsigned short \f2uled\f4)\f1
.br
\f4unsigned short v470_read_mode(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_write_mode(V470_REGS *\f2pb\f4, unsigned short \f2mode\f4)\f1
.br
\f4unsigned short v470_read_calid(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_read_biss(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_read_ycal(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_read_dcal(V470_REGS *\f2pb\f4)\f1
.br
\f4unsigned short v470_read_bistblk(V470_REGS *\f2pb\f4, int \f2blknum\f4, unsigned short *\f2bistblk\f4)\f1 // bist block needs to point to an array of 6 unsigned short
.br
\f4char *v470_get_cal_date(V470_REGS *\f2pb\f4)\f1
.br
\f4int v470_read_reg(V470_REGS *\f2pb\f4, unsigned short \f2offset\f4, unsigned short *\f2reg_value\f4)\f1
.br
\f4int v470_read_interlock_reg(V470_REGS *\f2pb\f4, unsigned short \f2offset\f4, unsigned int *\f2reg_value\f4)\f1
.br

.SS MACRO EXECUTION
\f4int v470_execute_macro(V470_REGS *\f2pb\f4, unsigned short \f2macro\f4, unsigned short *\f2params\f4, unsigned short *\f2arg\f4)\f1
.br

.SS CONFIGURATION
\f4int v470_config_rtd_type(V470_REGS *\f2pb\f4, char \f2rtd_id\f4, int \f2type\f4)\f1
.br
\f4int v470_config_analog_chnl(V470_REGS *\f2pb\f4, unsigned int \f2chnl_num\f4, unsigned int \f2rn_code\f4)\f1
.br
\f4int v470_config_tc_chnl(V470_REGS *\f2pb\f4, unsigned int \f2chnl_num\f4, unsigned int \f2rs_code\f4, unsigned int \f2rn_code\f4)\f1
.br

.SS READ/WRITE OPERATIONS
\f4int v470_read_rtd_type(V470_REGS *\f2pb\f4, char \f2rtd_id\f4, int *\f2ptype\f4)\f1
.br
\f4int v470_read_rtd_temp(V470_REGS *\f2pb\f4, char \f2rtd_id\f4, char \f2deg_spec\f4, float *\f2ptemp\f4)\f1
.br
\f4int v470_read_rtd_ohms(V470_REGS *\f2pb\f4, char \f2rtd_id\f4, float *\f2pohms\f4)\f1
.br
\f4int v470_write_fake_rtd_temp(V470_REGS *\f2pb\f4, char \f2rtd_id\f4, char \f2deg_spec\f4, float \f2temp\f4)\f1
.br
\f4int v470_write_analog_chnl(V470_REGS *\f2pb\f4, int \f2chnl_num\f4, float \f2volts_data\f4)\f1
.br
\f4int v470_write_tc_chnl(V470_REGS *\f2pb\f4, int \f2chnl_num\f4, char \f2deg_spec\f4, float \f2temp\f4)\f1
.br
\f4int v470_chnl_to_rn_code(V470_REGS *\f2pb\f4, int \f2chnl\f4, unsigned short *\f2prn_code\f4)\f1
.br
\f4int v470_rn_code_to_range(int \f2rn_code\f4, float *\f2pmin\f4, float *\f2pmax\f4)\f1
.br

.SH DESCRIPTION
The \f4v470_util\f1 library provides an easy method for a user to address functions of the Highland Technology V470 digital to
analog converter board.  The first argument \f2pb\f1 of these routines is a virtual address pointer to the base address of the
board.  This can be attained through the use of the \f4v120_config(1)\f1 and \f4get_vaddr_for_device(3)\f1 interface, or through
\f4v120_open(3)\f1, \f4v120_add_vme_region(3)\f1, and \f4v120_allocate_vme(3)\f1.

.TP
\f4#include \f2"v470_utils_proto.h"\f1
.br
defines the prototypes for all of these calls.
.TP
\f4#include \f2"v470_regs_sdef.h"\f1
.br
defines all the register #define constants, and the register map of the board.

.SS \f4unsigned short v470_read_vximfr(V470_REGS *\f2pb\f4)\f1
Returns the value of the VXIMFR register.  Highland Technology's registered code is 0xfeee.

.SS \f4unsigned short v470_read_vxitype(V470_REGS *\f2pb\f4)\f1
Returns the value of the VXITYPE register.  These are defined in \f2vme_search_specs_stor.h\f1.

.nf
.in +5
v210 -- VME Relay Module                                              0x56c2
v220 -- VME 4-20mA Analog Control I/O Module                          0x56cc
v230 -- VME Analog Input Module                                       0x56d6
v250 -- VME Digital I/O Module                                        0x56ea
v340 -- VME Waveform Generator Module                                 0x5744
v344 -- VME Waveform Generator                                        0x5648
v346 -- VME Waveform Generator                                        0x564a
v350 -- VME Analog/Digital Function Generator                         0x574e
v360 -- VME Tachometer Module                                         0x5758
v365 -- VME Tachometer / Overspeed Module                             0x575d
v370 -- VME Waveform Generator                                        0x5762
v375 -- VME Waveform Generator                                        0x5767
v380 -- VME Strain Gauge / Weighing Module                            0x576c
v385 -- VME Strain Gauge / Load Cell Module                           0x5771
v410 -- VME RTD / Resistance Input Module                             0x578a
v420 -- VME Isolated Resistance Simulator                             0x5794
v450 -- VME Analog Input Module                                       0x57b2
v460 -- VME Analog Scanner Module                                     0x57bc
v470 -- VME Analog Output and Thermocouple Simulator Module           0x57c6
v490 -- VME Multi-range Digitizer                                     0x57da
v545 -- VME Synchro/Resolver/LVDT/RVDT Simulation/Acquisition Module  0x5811
.in -5
.fi

.SS \f4unsigned short v470_read_serial(V470_REGS *\f2pb\f4)\f1
Returns the value of the SERIAL register.

.SS \f4unsigned short v470_read_romid(V470_REGS *\f2pb\f4)\f1
Returns the value of the ROMID register.

.SS \f4unsigned short v470_read_romrev(V470_REGS *\f2pb\f4)\f1
Returns the value of the ROMREV register.

.SS \f4unsigned short v470_read_cflags(V470_REGS *\f2pb\f4)\f1
Returns the value of the CFLAGS register.

.SS \f4unsigned short v470_read_rflags(V470_REGS *\f2pb\f4)\f1
Returns the value of the RFLAGS register.

.SS \f4unsigned short v470_read_relays(V470_REGS *\f2pb\f4)\f1
Returns the value of the RELAYS register.

.SS \f4unsigned short v470_write_relays(V470_REGS *\f2pb\f4, unsigned short \f2relays\f4)\f1
Writes the \f2relays\f1 into the RELAYS register -- controls calibration bus relays.

.SS \f4unsigned short v470_read_uled(V470_REGS *\f2pb\f4)\f1
Returns the value of the ULED register.

.SS \f4unsigned short v470_write_uled(V470_REGS *\f2pb\f4, unsigned short \f2uled\f4)\f1
Writes the value of \f2uled\f1 into the ULED register.

.SS \f4unsigned short v470_read_mode(V470_REGS *\f2pb\f4)\f1
Returns the value of the MODE register.

.SS \f4unsigned short v470_write_mode(V470_REGS *\f2pb\f4, unsigned short \f2mode\f4)\f1
Writes the value of \f2mode\f1 into the MODE register.

.SS \f4unsigned short v470_read_calid(V470_REGS *\f2pb\f4)\f1
Returns the value of the CALID register.

.SS \f4unsigned short v470_read_biss(V470_REGS *\f2pb\f4)\f1
Returns the value of the BISS register.

.SS \f4unsigned short v470_read_ycal(V470_REGS *\f2pb\f4)\f1
Returns the value of the YCAL register -- calibration date -- year.

.SS \f4unsigned short v470_read_dcal(V470_REGS *\f2pb\f4)\f1
Returns the value of the DCAL register -- calibration date -- MM/DD.

.SS \f4unsigned short v470_read_bistblk(V470_REGS *\f2pb\f4, int \f2blknum\f4, unsigned short *\f2bistblk\f4)\f1
Returns an array of 6 BIST block words to 6 word array pointed to by \f2bistblk\f1.  Errors are returned in six word blocks
in registers 128-255, for a total of 21 total errors.  \f2blknum\f1 specifies which of the 21 blocks to load.

.SS \f4char *v470_get_cal_date(V470_REGS *\f2pb\f4)\f1
Returns calibration date as a string in the format of "MM/DD/YYYY".

.SS \f4int v470_read_reg(V470_REGS *\f2pb\f4, unsigned short \f2offset\f4, unsigned short *\f2reg_value\f4)\f1
Returns value of register at \f2offset\f1.

.SS \f4int v470_read_interlock_reg(V470_REGS *\f2pb\f4, unsigned short \f2offset\f4, unsigned int *\f2reg_value\f4)\f1
Returns 32-bit value of interlocked registers at offset 44-55.

Example:

Read channel loopback voltage from BIVH:BIVL registers:

.nf
.in +5
#define LSB 0.00000000745058L

double readbackvolts;
int bistloopbackcounts;

if (v470_read_interlock_reg(pb, 0x6c, (unsigned int *)&bistloopbackcounts) == -1)
  {
    ... print error message ...
  }
.in -5
.fi

readbackvolts = (double)bistloopbackcounts * LSB * 16.0L

.SS \f4int \f2v470_execute_macro(V470_REGS *\f2pb\f4, unsigned short \f2macro\f4, unsigned short *\f2params\f4, unsigned short *\f2arg\f4)\f1
.TP
\f2pb\f1
.br
virtual address pointer to base of board.
.TP
\f2macro\f1
.br
supports the following macros as defined in the manual and \f2v470_regs_sdef.h\f1.

.nf
.in +5
#define V470_MACRO_NOOP                 0x0000
#define V470_MACRO_SETALL_TYPE_J        0x8401  // set all channels to Type J thermocouple at 16.7 Hz, using the onboard reference junction
#define V470_MACRO_SETALL_TYPE_K        0x8402  // set all channels to Type K thermocouple at 16.7 Hz, using the onboard reference junction
#define V470_MACRO_SETALL_TYPE_E        0x8403  // set all channels to Type E thermocouple at 16.7 Hz, using the onboard reference junction
#define V470_MACRO_SETALL_TYPE_T        0x8404  // set all channels to Type T thermocouple at 16.7 Hz, using the onboard reference junction
#define V470_MACRO_SETALL_12_5_V        0x8405  // set all channels to +/- 12.5V output
#define V470_MACRO_SETALL_80_MV         0x8406  // set all channels to +/- 80mV
#define V470_MACRO_SETALL_25_MV         0x8407  // set all channels to +/- 25VmV
#define V470_MACRO_BIST_ALL             0x8410  // BIST all
#define V470_MACRO_BIST_ONE_CHNL        0x8411  // BIST one channel
#define V470_MACRO_HARD_REBOOT          0x8420  // hard reboot -- reloads PFGSs, restarts code, disappears from bus for 4 seconds
#define V470_MACRO_SOFT_REBOOT          0x8421  // soft reboot -- remains on bus
.in -5
.fi

.TP
\f2params\f1
.br
used as a value or bitmask in the following macros:

.nf
.in +5
V470_MACRO_BIST_ONE_CHNL
.in -5
.fi

.TP
\f2arg\f1
.br
presently unused -- just enter 0.

Example 1:

BIST channel 2.

.nf
.in +5
unsigned short param0 = 2;

if (v470_execute_macro(pb, V470_MACRO_BIST_ONE_CHNL, &param0, 0) == -1)
  {
    ... print error message ...
  }
.in -5
.fi

Example 2:

Reset the board.

.nf
.in +5
if (v470_execute_macro(pb, V470_MACRO_HARD_REBOOT, 0, 0) == -1)
  {
    ... print error message ...
    exit(1);
  }
.in -5
.fi

.SS \f4int v470_config_rtd_type(V470_REGS *\f2pb\f4, char \f2rtd_id\f4, int \f2type\f4)\f1
Sets the type of the RTD -- either 100 or 1000 ohm platinum.
.TP
\f2pb\f1
.br
virtual address pointer to base of board.
.TP
\f2rtd_id\f1
.br
is 'a', 'b', 'c', or 'd'.
.TP
\f2type\f1
.br
is one of the following (from v470_utils/v470_regs_sdef.h):

.nf
.in +5
// RTD types

#define V470_RTD_TYPE_UNUSED            0
#define V470_RTD_TYPE_100_OHM_PT        1
#define V470_RTD_TYPE_1000_OHM_PT       2
.in -5
.fi

.SS \f4int v470_config_analog_chnl(V470_REGS *\f2pb\f4, unsigned int \f2chnl_num\f4, unsigned int \f2rn_code\f4)\f1
Configures the analog channel range.
.TP
\f2pb\f1
.br
virtual address pointer to base of board.
.TP
\f2chnl_num\f1
.br
the valid range for \f2chnl_num\f1 is 0-15.
.TP
\f2rn_code\f1.
.br
is one of the following (from v470_utils/v470_regs_sdef.h):

.nf
.in +5
#define V470_RN_CODE_OFF                0
#define V470_RN_CODE_25_MV              1       // +/-25mv
#define V470_RN_CODE_50_MV              2       // +/-50mv
#define V470_RN_CODE_80_MV              3       // +/-80mv
#define V470_RN_CODE_125_MV             4       // +/-125mv
#define V470_RN_CODE_250_MV             5       // +/-250mv
#define V470_RN_CODE_500_MV             6       // +/-500mv
#define V470_RN_CODE_1_25_V             7       // +/-1.25v
#define V470_RN_CODE_2_5_V              8       // +/-2.5v

lib/v470_utils/v470_utils.c
// -*- c++ -*-

// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or (at
// your option) any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
//
// Author: Dean W. Anneser
// Company: RTLinux Solutions LLC for Highland Technology, Inc.
// Date: Sun Aug 08 11:43:00 2021

#include <stdio.h>
#include <string.h>
#include <unistd.h>

// prototypes

#include "deg_conv_proto.h"
#include "ht_utils_proto.h"
#include "misc_math_utils_proto.h"
#include "v470_utils_proto.h"

// structure defs

#include "v470_regs_sdef.h"

unsigned short v470_read_vximfr(V470_REGS *pb)
{
  return(pb->vximfr);
}


unsigned short v470_read_vxitype(V470_REGS *pb)
{
  return(pb->vxitype);
}


unsigned short v470_read_serial(V470_REGS *pb)
{
  return(pb->serial);
}


unsigned short v470_read_romid(V470_REGS *pb)
{
  return(pb->romid);
}


unsigned short v470_read_romrev(V470_REGS *pb)
{
  return(pb->romrev);
}


unsigned short v470_read_cflags(V470_REGS *pb)
{
  return(pb->cflags);
}


unsigned short v470_read_rflags(V470_REGS *pb)
{
  return(pb->rflags);
}


unsigned short v470_read_relays(V470_REGS *pb)
{
  return(pb->relays);
}


unsigned short v470_write_relays(V470_REGS *pb, unsigned short relays)
{
  pb->relays = relays;
  return(0);
}


unsigned short v470_read_uled(V470_REGS *pb)
{
  return(pb->uled);
}


unsigned short v470_write_uled(V470_REGS *pb, unsigned short uled)
{
  pb->uled = uled;
  return(0);
}


unsigned short v470_read_mode(V470_REGS *pb)
{
  return(pb->mode);
}


unsigned short v470_write_mode(V470_REGS *pb, unsigned short mode)
{
  pb->mode = mode;
  return(0);
}


unsigned short v470_read_calid(V470_REGS *pb)
{
  return(pb->calid);
}


unsigned short v470_read_biss(V470_REGS *pb)
{
  return(pb->biss);
}


unsigned short v470_read_ycal(V470_REGS *pb)
{
  return(pb->ycal);
}


unsigned short v470_read_dcal(V470_REGS *pb)
{
  return(pb->dcal);
}


unsigned short v470_read_bistblk(V470_REGS *pb, int blknum, unsigned short *bistblk) // bist block needs to point to an array of 6 unsigned short
{
  int i;
  unsigned short *pbistblk = (unsigned short *)&pb->buffer[6 * blknum];
  unsigned short *puserbistblk = bistblk;

  for (i = 0; i < 6; i++)
    *puserbistblk++ = *pbistblk++;

  return(0);
}


char *v470_get_cal_date(V470_REGS *pb)
{
  static char date_string[16];
  int year, month, day;

  memset((void *)date_string, 0, 16);
  year = (int)v470_read_ycal(pb);
  month = (int)v470_read_dcal(pb) >> 8;
  day = (int)v470_read_dcal(pb) & 0xff;
  sprintf(date_string, "%d/%d/%d", month, day, year);
  return(date_string);
}

int v470_read_reg(V470_REGS *pb, unsigned short offset, unsigned short *reg_value)
{
  *reg_value = *(unsigned short *)((void *)pb + (offset & 0xfffe));
  return(0);
}


int v470_read_interlock_reg(V470_REGS *pb, unsigned short offset, unsigned int *reg_value)
{
  *reg_value = ht_read_interlocked((unsigned short *)((void *)pb + (offset & 0xfffc)));
  return(0);
}


// execute macros

int v470_execute_macro(V470_REGS *pb, unsigned short macro, unsigned short *params, unsigned short *arg)
{
  int rtc = 0;
  switch(macro)
    {
    case V470_MACRO_NOOP:
      return(0);

    case V470_MACRO_SETALL_TYPE_J:	// set all channels to Type J thermocouple at 16.7 Hz, using the onboard reference junction
      return(ht_write_macro((unsigned short *)&pb->macro, V470_MACRO_SETALL_TYPE_J, 0, 0, 0));	// should complete in 1 msec

    case V470_MACRO_SETALL_TYPE_K:	// set all channels to Type K thermocouple at 16.7 Hz, using the onboard reference junction
      return(ht_write_macro((unsigned short *)&pb->macro, V470_MACRO_SETALL_TYPE_K, 0, 0, 0));	// should complete in 1 msec

    case V470_MACRO_SETALL_TYPE_E:	// set all channels to Type E thermocouple at 16.7 Hz, using the onboard reference junction
      return(ht_write_macro((unsigned short *)&pb->macro, V470_MACRO_SETALL_TYPE_E, 0, 0, 0));	// should complete in 1 msec

    case V470_MACRO_SETALL_TYPE_T:	// set all channels to Type T thermocouple at 16.7 Hz, using the onboard reference junction
      return(ht_write_macro((unsigned short *)&pb->macro, V470_MACRO_SETALL_TYPE_T, 0, 0, 0)); 	// should complete in 1 msec

    case V470_MACRO_SETALL_12_5_V:	// set all channels to +/- 12.5V at 16.7 Hz
      return(ht_write_macro((unsigned short *)&pb->macro, V470_MACRO_SETALL_12_5_V, 0, 0, 0));	// should complete in 1 msec

    case V470_MACRO_SETALL_80_MV:		// set all channels to +/- 80V at 16.7 Hz
      return(ht_write_macro((unsigned short *)&pb->macro, V470_MACRO_SETALL_80_MV, 0, 0, 0));	// should complete in 1 msec

    case V470_MACRO_SETALL_25_MV:		// set all channels to +/- 25V at 16.7 Hz
      return(ht_write_macro((unsigned short *)&pb->macro, V470_MACRO_SETALL_25_MV, 0, 0, 0));	// should complete in 1 msec

    case V470_MACRO_BIST_ALL:   		// test all
      return(ht_write_macro((unsigned short *)&pb->macro, V470_MACRO_BIST_ALL, 0, 0, 0));	// should complete in 80 sec
             
    case V470_MACRO_BIST_ONE_CHNL:
      return(ht_write_macro((unsigned short *)&pb->macro, V470_MACRO_BIST_ONE_CHNL, params, 1, 0));	// should complete in 5 sec

    case V470_MACRO_HARD_REBOOT:	// hard reboot -- reloads PFGSs, restarts code, disappears from bus for 4 seconds
      rtc = ht_write_macro((unsigned short *)&pb->macro, V470_MACRO_HARD_REBOOT, 0, 0, 1);	// should complete in 5 secs
      sleep(5);
      return(rtc);

    case V470_MACRO_SOFT_REBOOT:	// soft reboot -- remains on bus
      rtc = ht_write_macro((unsigned short *)&pb->macro, V470_MACRO_SOFT_REBOOT, 0, 0, 1);	// should complete in 1 sec
      sleep(2);
      return(rtc);

    default:
      return(-1);
    }
}


// v470_config_rtd_type(pb, 'a', V470_RTD_TYPE_100_OHM_PT)		# set RTD a to 100 ohm PT

int v470_config_rtd_type(V470_REGS *pb, char rtd_id, int type)
{
  switch(rtd_id)
    {
    case 'a':
      switch(type)
        {
        case '0':
          pb->rtda = V470_RTD_TYPE_UNUSED;
          break;
        case 1:
          pb->rtda = V470_RTD_TYPE_100_OHM_PT; // set to 100 ohm platinum
          break;
        case 2:
          pb->rtda = V470_RTD_TYPE_1000_OHM_PT; // set to 1000 ohm platinum
          break;
        default:
          return(-1);
        }
      break;

    case 'b':
      switch(type)
        {
        case '0':
          pb->rtdb = V470_RTD_TYPE_UNUSED;
          break;
        case 1:
          pb->rtdb = V470_RTD_TYPE_100_OHM_PT; // set to 100 ohm platinum
          break;
        case 2:
          pb->rtdb = V470_RTD_TYPE_1000_OHM_PT; // set to 1000 ohm platinum
          break;
        default:
          return(-1);
        }
      break;

    case 'c':
      switch(type)
        {
        case '0':
          pb->rtdc = V470_RTD_TYPE_UNUSED;
          break;
        case 1:
          pb->rtdc = V470_RTD_TYPE_100_OHM_PT; // set to 100 ohm platinum
          break;
        case 2:
          pb->rtdc = V470_RTD_TYPE_1000_OHM_PT; // set to 1000 ohm platinum
          break;
        default:
          return(-1);
        }
      break;

    case 'd':
      switch(type)
        {
        case '0':
          pb->rtdd = V470_RTD_TYPE_UNUSED;
          break;
        case 1:
          pb->rtdd = V470_RTD_TYPE_100_OHM_PT; // set to 100 ohm platinum
          break;
        case 2:
          pb->rtdd = V470_RTD_TYPE_1000_OHM_PT; // set to 1000 ohm platinum
          break;
        default:
          return(-1);
        }
      break;

    default:
      return(-1);
    }
  return(0);
}


// v470_config_analog_chnl(pb, chnl_num, rn_code)

int v470_config_analog_chnl(V470_REGS *pb, unsigned int chnl_num, unsigned int rn_code)
{
  unsigned short chnl_ctl = 0;

  chnl_ctl = ((unsigned short)rn_code << V470_RN_CODE_SHIFT);
  pb->ccb[chnl_num].ctl = chnl_ctl;

  return(0);
}


// v470_config_tc_chnl(pb, chnl_num, rs_code, rn_code)

int v470_config_tc_chnl(V470_REGS *pb, unsigned int chnl_num, unsigned int rs_code, unsigned int rn_code)
{
  unsigned short chnl_ctl = 0;

  chnl_ctl = (((unsigned short)rn_code << V470_RN_CODE_SHIFT) | ((unsigned short)rs_code << V470_RS_CODE_SHIFT));
  pb->ccb[chnl_num].ctl = chnl_ctl;
  
  return(0);
}


// v470_read_rtd_type(pb, 'b', &type)		# read RTD b type (a-d) and load into parameter type (0 = unused, 1 = 100 ohm PT, or 2 = 1000 ohm PT)

int v470_read_rtd_type(V470_REGS *pb, char rtd_id, int *ptype)
{
  switch(rtd_id)
    {
    case 'a':
      *ptype = (int)pb->rtda;
      break;

    case 'b':
      *ptype = (int)pb->rtdb;
      break;

    case 'c':
      *ptype = (int)pb->rtdc;
      break;

    case 'd':
      *ptype = (int)pb->rtdd;
      break;

    default:
      return(-1);
    }

  return(0);
}


// v470_read_rtd_temp(pb, 'r', 'F', &temp)	# read reference (internal) temp (a-d,r,1,2), convert to Fahrenheit, and load into parameter temp

int v470_read_rtd_temp(V470_REGS *pb, char rtd_id, char deg_spec, float *ptemp)
{
  float deg;
  float degC;

  switch(rtd_id)
    {
    case 'a':
      degC = (float)pb->tempa / 16.0;
      break;

    case 'b':
      degC = (float)pb->tempb / 16.0;
      break;

    case 'c':
      degC = (float)pb->tempc / 16.0;
      break;

    case 'd':
      degC = (float)pb->tempd / 16.0;
      break;

    case 'r':
      degC = (float)pb->tempr / 16.0;
      break;

    case '1':
      degC = (float)pb->fake1 / 16.0;
      break;

    case '2':
      degC = (float)pb->fake2 / 16.0;
      break;

    default:
      return(-1);
    }

  switch(deg_spec)
    {
    case 'C':
      deg = degC;
      break;
    case 'F':
      deg = degC_to_degF(degC);
      break;
    case 'K':
      deg = degC_to_degK(degC);
      break;
    case 'R':
      deg = degC_to_degR(degC);
      break;
    default:
      return(-1);
    }
  
  *ptemp = deg;
  return(0);
}


// v470_read_rtd_ohms(pb, 't', &ohms)		# read internal test resistor resistance in ohms, and load into parameter ohms

int v470_read_rtd_ohms(V470_REGS *pb, char rtd_id, float *pohms)
{
  switch(rtd_id)
    {
    case 'a':
      *pohms = (float)ht_read_interlocked(&pb->rahi) / 65536.0;
      break;

    case 'b':
      *pohms = (float)ht_read_interlocked(&pb->rbhi) / 65536.0;
      break;

    case 'c':
      *pohms = (float)ht_read_interlocked(&pb->rchi) / 65536.0;
      break;

    case 'd':
      *pohms = (float)ht_read_interlocked(&pb->rdhi) / 65536.0;
      break;

    case 't':
      *pohms = (float)ht_read_interlocked(&pb->trhi) / 65536.0;
      break;
      
    default:
      return(-1);
    }

  return(0);
}


// v470_write_fake_rtd_temp(pb, '1', 'C', 100.0)	# write 100 degC into RTD fake 1

int v470_write_fake_rtd_temp(V470_REGS *pb, char rtd_id, char deg_spec, float temp)
{
  float degC;
  
  switch(deg_spec)
    {
    case 'C':
      degC = temp;
      break;
    case 'F':
      degC = degF_to_degC(temp);
      break;
    case 'K':
      degC = degK_to_degC(temp);
      break;
    case 'R':
      degC = degR_to_degC(temp);
      break;
    default:
      return(-1);
    }
  
  switch(rtd_id)
    {
    case '1':
      pb->fake1 = degC * 16.0;
      break;
    case '2':
      pb->fake2 = degC * 16.0;
      break;
    default:
      return(-1);
    }
  return(0);
}


// v470_write_analog_chnl(pb, chnl_num, &volts_data)	# write volts 

int v470_write_analog_chnl(V470_REGS *pb, int chnl_num, float volts_data)
{
  unsigned short rn_code;
  float vmin, vmax;
  double counts;


  if (v470_chnl_to_rn_code(pb, chnl_num, &rn_code) == -1)
    return(-1);

  if (v470_rn_code_to_range(rn_code, &vmin, &vmax) == -1)
    return(-1);

  counts = clamp(ratio((double)volts_data, vmin, vmax, V470_COUNTS_MIN_16, V470_COUNTS_MAX_16), V470_COUNTS_MIN_16, V470_COUNTS_MAX_16 - 1);
  pb->ccb[chnl_num].val = (unsigned short)counts;
  return(0);
}


// v470_write_tc_chnl(pb, chnl_num, deg_spec, temp)

int v470_write_tc_chnl(V470_REGS *pb, int chnl_num, char deg_spec, float temp)
{
  float degC;
  unsigned short rn_code;
  float tmin, tmax;

  switch(deg_spec)
    {
    case 'C':
      degC = temp;
      break;
    case 'F':
      degC = degF_to_degC(temp);
      break;
    case 'K':
      degC = degK_to_degC(temp);
      break;
    case 'R':
      degC = degR_to_degC(temp);
      break;
    default:
      return(-1);
    }
  
  if (v470_chnl_to_rn_code(pb, chnl_num, &rn_code) == -1)
    return(-1);

  if (v470_rn_code_to_range(rn_code, &tmin, &tmax) == -1)
    return(-1);

  degC = (float)clamp((double)degC, (double)tmin, (double)tmax);
  pb->ccb[chnl_num].val = (short)(degC * 16.0);
  return(0);
}


int v470_chnl_to_rn_code(V470_REGS *pb, int chnl, unsigned short *prn_code)
{
  *prn_code = (pb->ccb[chnl].ctl & V470_RN_CODE_MASK) >> V470_RN_CODE_SHIFT;
  return(0);
}


int v470_rn_code_to_range(int rn_code, float *pmin, float *pmax)
{
  if ((pmin == (float *)0) || (pmax == (float *)0))
    return(-1);

  switch(rn_code)
    {
    case V470_RN_CODE_OFF:
      *pmin = 0.0;
      *pmax = 0.0;
      break;
    case V470_RN_CODE_25_MV:
      *pmin = -0.025;
      *pmax =  0.025;
      break;
    case V470_RN_CODE_50_MV:
      *pmin = -0.050;
      *pmax =  0.050;
      break;
    case V470_RN_CODE_80_MV:
      *pmin = -0.080;
      *pmax =  0.080;
      break;
    case V470_RN_CODE_125_MV:
      *pmin = -0.125;
      *pmax =  0.125;
      break;
    case V470_RN_CODE_250_MV:
      *pmin = -0.25;
      *pmax =  0.25;
      break;
    case V470_RN_CODE_500_MV:
      *pmin = -0.5;
      *pmax =  0.5;
      break;
    case V470_RN_CODE_1_25_V:
      *pmin = -1.25;
      *pmax =  1.25;
      break;
    case V470_RN_CODE_2_5_V:
      *pmin = -2.5;
      *pmax =  2.5;
      break;
    case V470_RN_CODE_5_V:
      *pmin = -5.0;
      *pmax =  5.0;
      break;
    case V470_RN_CODE_12_5_V:
      *pmin = -12.5;
      *pmax =  12.5;
      break;
    case V470_RN_CODE_TYPE_J:		// -210 to 1200 degC
      *pmin = -210.0;
      *pmax = 1200.0;
      break;
    case V470_RN_CODE_TYPE_K:		// -270 to 1372 degC
      *pmin = -270.0;
      *pmax = 1372.0;
      break;
    case V470_RN_CODE_TYPE_E:		// -270 to 1000 degC
      *pmin = -270.0;
      *pmax = 1000.0;
      break;
    case V470_RN_CODE_TYPE_T:		// -270 to 400 degC
      *pmin = -270.0;
      *pmax = 400.0;
      break;
    case V470_RN_CODE_TYPE_R:		// -50 to 1768 degC
      *pmin = -50.0;
      *pmax = 1768.0;
      break;
    case V470_RN_CODE_TYPE_S:		// -50 to 1768 degC
      *pmin = -50.0;
      *pmax = 1768.0;
      break;
    case V470_RN_CODE_TYPE_B:		// 0 to 1820 degC
      *pmin = 0.0;
      *pmax = 1820.0;
      break;
    case V470_RN_CODE_TYPE_N:		// -270 to 1300 degC
      *pmin = -270.0;
      *pmax = 1300.0;
      break;
    default:
      return(-1);
    }
  return(0);
}


